File: cpp/316. [M] Remove Duplicate Letters.cpp
#include <vector>
#include <unordered_map>
#include <assert.h>
#include <iostream>
#include <list>
#include <stack>
#include <unordered_set>

using namespace std;;

class Solution {
public:
    string removeDuplicateLetters(string s) {
        stack<char> stk;
        unordered_map<char, int> m;
        unordered_set<char> set;

        for (char c : s) {
            if (m.find(c) == m.end()) {
                m[c] = 0;
            }
            m[c]++;
        }

        for (char c : s) {
            // cout << c << "00000" << endl;
            if (set.find(c) != set.end()) {
                m[c]--;
                continue;
            }
            // cout << c << "11111" << endl;
            while (!stk.empty() && stk.top() >= c && m[stk.top()]) {
                cout << stk.top() << m[stk.top()] << endl;
                set.erase(stk.top());
                stk.pop();
            }
            stk.push(c);
            set.insert(c);
            m[c]--;
        }

        string res = "";
        while (!stk.empty()) {
            res = stk.top() + res;
            stk.pop();
        }
        return res;
    }
};

int main() {
    Solution sol = Solution();
    vector<int> input = {1,2,3,4,5,6,7};
    cout << sol.removeDuplicateLetters("bcabc") << endl;
    cout << sol.removeDuplicateLetters("cbacdcbc") << endl;
    cout << sol.removeDuplicateLetters("bbcaac") << endl;
}

File: python/316. [M] Remove Duplicate Letters.py
import heapq


class Solution:

    def removeDuplicateLetters(self, s: str) -> str:
        c_to_last_i = {}
        for i, c in enumerate(s):
            c_to_last_i[c] = i
        stack = []
        for i, char in enumerate(s):
            if char in stack:
                continue
            while stack and c_to_last_i[stack[-1]] > i and char < stack[-1]:
                stack.pop()
            stack.append(char)
        return ''.join(stack)

    def removeDuplicateLetters_version0(self, s: str) -> str:
        c_to_last_i = {}
        for i, c in enumerate(s):
            c_to_last_i[c] = i
        last_chars = sorted([(val, key) for key, val in c_to_last_i.items()])
        res = ''
        start_i = 0
        # print(last_chars)
        for last_i, char in last_chars:
            if char in res:
                continue
            while start_i <= last_i and char not in res:
                sub_str = s[start_i:last_i + 1]
                # print(sub_str, '---')
                heap = list(sub_str)
                heapq.heapify(heap)
                while heap and heap[0] in res:
                    heapq.heappop(heap)
                if not heap:
                    start_i = last_i + 1
                    break
                sub_str_min_char = heap[0]
                sub_str_min_char_i = sub_str.index(sub_str_min_char) + start_i
                res += sub_str_min_char
                start_i = sub_str_min_char_i + 1
                # print(res)
        return res


if __name__ == '__main__':
    sol = Solution()
    print(sol.removeDuplicateLetters('abaacbbca'))
    print(sol.removeDuplicateLetters('abadddacbddbc'))